src/charm.py

```diff
@@ -649,13 +726,29 @@ def _on_update_status(self, _):
         # before an IP address was ready, leaving UpdateStatue as the last resort.
         self._common_exit_hook()
 
+    def _calculate_current_config_hash(self):
+        # When restoring the charm on upgrade, we need to calculate the stored config hash based on
+        #  the actual container fs contents.
+
+        def _normalize(raw_yaml_str):
+            # to remove possible whitespace inconsistencies, deserialize and reserialize
+            # todo: since we own the files and (should) control when they're
+            #  written, is this really necessary?
+            return yaml.safe_dump(yaml.safe_load(raw_yaml_str))
+
+        config_state = ConfigFileSystemState()
+        # add all configuration paths to the state
+        for path in self._config_paths:
+            content = _normalize(self.container.pull(self._config_path))
+            config_state.add_file(path, content)
+
+        return hash(config_state)
```

Comment:
We decided to avoid sub-functions a while ago...

Reply:
Meh :D